text/2018-04-08-global-and-local-pinning.md

```diff
@@ -0,0 +1,247 @@
+# Summary
+
+Introduce the concept of *global pinning*, which allows one to pin the current thread without going through a thread-local handle.
+
+In addition to that, we'll slightly tweak the general interface of `crossbeam-epoch` in order to fix several reported problems with it.
+
+# Motivation
+
+The following problems with the current `Collector`/`Handle`/`Guard` interface have been reported:
+
+1. Users might mistakenly expect `Handle::clone()` to register a brand-new handle rather than returning a reference to the same handle. The method can be dangerously misleading.
+
+2. Handles are not `Send`. This is an annoying problem and it's not obvious how to cheat around it correctly when you really have to.
+
+3. Handles are not `Sync`. Sometimes it'd be nice to call `pin()` without registering a handle for each thread. In `SkipMap`, we need to pin inside `self.collector`, but creating a handle per-thread *and* per-instance is difficult. In `elfmalloc` there are situations where having to register handles is a nuisance.
+
+4. `epoch::pin()` will panic if the thread-local handle is destroyed. Fortunately, `LocalKey::try_with` will be soon stabilized in Rust 1.26. But how do we work around the panics then - what should we do exactly in case the handle is destroyed?
+
+# Detailed design
+
+### Global pinning
+
+Let's introduce something I've been referring to as *anonymous pinning* in the past, and has been briefly explained in [this comment](https://github.com/crossbeam-rs/crossbeam-epoch/pull/21#discussion_r144353198).
+
+The idea is to have two atomic counters in the `Collector` (more concretely, in the `Global` struct). They count how many `Guards` are currently pinned in the global epoch and the epoch preceding it:
+
+```rust
+pin_counters: [AtomicUsize; 2],
+```
+
+We introduce a new method `Collector::pin()`, which is just like `Handle::pin()`:
+
+```rust
+impl Collector {
+    fn pin(&self) -> Guard;
+}
+```
+
+We'll have to distinguish between guards pinned globally and locally. Since we're turning `Guard` into an `enum`, let's also add a special case for the unprotected guard:
+
+```rust
+enum Guard {
+    Global(*const Global),
+    Local(*const Local),
+    Unprotected,
+}
+```
+
+Now, the `Global::pin()` method might be implemented like this:
```

Comment:
Did you mean `Collector::pin()`?

Reply:
Well, `Collector::pin()` would simply call `Global::pin()`, just like `Handle::pin()` now calls `Local::pin()`. It's the same thing :)

Reply:
(Oh, fsr I hadn't realized `Global` is also a type and not only a variant.)